src/Microsoft.OData.Edm/ExtensionMethods/EdmTypeSemantics.cs

```diff
@@ -1216,18 +1221,39 @@ public static EdmPrimitiveTypeKind PrimitiveKind(this IEdmTypeReference type)
         /// <param name="potentialBaseType">The potential base type of the type being tested.</param>
         /// <returns>True if and only if the type inherits from the potential base type.</returns>
         public static bool InheritsFrom(this IEdmStructuredType type, IEdmStructuredType potentialBaseType)
+        {            
+            string fullName = type.FullTypeName();
+            HashSet<IEdmStructuredType> baseTypes = GetBaseTypes(type);
+
+            return baseTypes.Contains(potentialBaseType);
+        }
+
+        private static HashSet<IEdmStructuredType> GetBaseTypes(IEdmStructuredType type)
         {
-            do
+            HashSet<IEdmStructuredType> baseTypes;
+
+            if (baseTypeCache.TryGetValue(type, out baseTypes))
             {
-                type = type.BaseType;
-                if (type != null && type.IsEquivalentTo(potentialBaseType))
+                return baseTypes;
+            }
+
+            baseTypes = new HashSet<IEdmStructuredType>();

```

Comment:
>baseTypes = new HashSet<IEdmStructuredType>(); [](start = 12, length = 46)

If a type hasn't any  base type, why do we need to create a HashSet?

Reply:
so that we have something to put in the dictionary (in this case, an empty hashset)

---
In reply to: [493002182](https://github.com/OData/odata.net/pull/1888#discussion_r493002182) [](ancestors = 493002182)

Reply:
Maybe to use "null" directly? We don't need to create an empty hashet? 

---
In reply to: [493027004](https://github.com/OData/odata.net/pull/1888#discussion_r493027004) [](ancestors = 493027004,493002182)

Reply:
Make null so we don't need to call Contains at line 1228

---
In reply to: [493028268](https://github.com/OData/odata.net/pull/1888#discussion_r493028268) [](ancestors = 493028268,493027004,493002182)

Reply:
I would be okay with that, but then we might consider changing:

    HashSet<IEdmStructuredType> GetBaseTypes(IEdmStructuredType type)

to:

   bool TryGetBaseTypes(IEdmStructuredType type, out HashSet<IEdmStructuredType>) 

and return false if there are no base types.

Either way works, since it's a private method whichever is easier is fine with me, but if we exposed this publicly I would rather make the empty base type explicit rather than returning a null and having the caller have to check for null (or not, which would likely lead to unexpected null ref exceptions)

---
In reply to: [493029669](https://github.com/OData/odata.net/pull/1888#discussion_r493029669) [](ancestors = 493029669,493028268,493027004,493002182)